markovflow/kalman_filter.py

```diff
@@ -495,3 +494,120 @@ def _log_det_observation_precision(self):
     def observations(self):
         """ Observation vector """
         return self.sites.means
+
+
+@tf_scope_class_decorator
+class KalmanFilterWithSparseSites(BaseKalmanFilter):
+    r"""
+    Performs a Kalman filter on a :class:`~markovflow.state_space_model.StateSpaceModel`
+    and :class:`~markovflow.emission_model.EmissionModel`, with Gaussian sites, over a time grid.
+    """
+
+    def __init__(self, state_space_model: StateSpaceModel, emission_model: EmissionModel, sites: GaussianSites,
+                 time_grid: tf.Tensor, observations_idx: tf.Tensor, observations: tf.Tensor):
+        """
+        :param state_space_model: Parameterises the latent chain.
+        :param emission_model: Maps the latent chain to the observations.
+        :param sites: Gaussian sites over the time grid.
+        :param time_grid: Time grid with shape (T,).
+        :param observations_idx: Index of the observations in the time grid with shape (N,).
+        :param observations: Observations with shape (N, state_dim).
+        """
+        self.sites = sites
+        self.time_grid = tf.expand_dims(time_grid, axis=-1)
+        self.observations_idx = observations_idx
+        self._observations = tf.scatter_nd(self.observations_idx, observations, self.time_grid.shape)
+        self.data_sites = self._get_data_sites()
+        super().__init__(state_space_model, emission_model)
+
+    def _get_data_sites(self):
+        """
+        Get data sites from all sites using observation idx.
+        """
+        nat1 = tf.gather_nd(self.sites.nat1, self.observations_idx)
+        nat2 = tf.gather_nd(self.sites.nat2, self.observations_idx)
+        log_norm = tf.gather_nd(self.sites.log_norm, self.observations_idx)
+
+        return UnivariateGaussianSitesNat(
+            nat1=nat1,
+            nat2=nat2,
+            log_norm=log_norm,
+        )
+
+    @property
+    def _r_inv(self):
+        """
+        Precisions of the observation model over the time grid.
+        """
+        data_sites_precision = self.data_sites.precisions
+        return tf.scatter_nd(self.observations_idx , data_sites_precision, self.time_grid[..., None].shape)
+
+    @property
+    def _log_det_observation_precision(self):
+        """
+        Sum of log determinant of the precisions of the observation model. It only calculates for the data_sites as
+        other sites precision is anyways zero.
+        """
+        data_r_inv = self._r_inv_data
+        return tf.reduce_sum(tf.linalg.logdet(data_r_inv), axis=-1)
```

Comment:
```suggestion
        return tf.reduce_sum(tf.linalg.logdet(self.data_sites.precisions), axis=-1)
```
I don't think we need that r_inv_data

Reply:
why not, it is actually used twice

Reply:
It's a function that does `self.data_sites.precisions` but yes that variable wasn't required so removed.